File: trunk/proj/Android/src/com/jasonzqshen/familyaccounting/core/transaction/MonthIdentity.java
package com.jasonzqshen.familyaccounting.core.transaction;

import java.io.Serializable;

import com.jasonzqshen.familyaccounting.core.exception.FiscalMonthRangeException;
import com.jasonzqshen.familyaccounting.core.exception.FiscalYearRangeException;
import com.jasonzqshen.familyaccounting.core.exception.MonthIdentityFormatException;

public class MonthIdentity implements Serializable, Comparable<MonthIdentity> {
	/**
	 * 
	 */
	private static final long serialVersionUID = -1145199479960583527L;
	public final int _fiscalYear;
	public final int _fiscalMonth;

	public MonthIdentity(int fiscalYear, int fiscalMonth)
			throws FiscalYearRangeException, FiscalMonthRangeException {
		if (fiscalYear < 1000 || fiscalYear > 9999) {
			throw new FiscalYearRangeException(fiscalYear);
		}
		if (fiscalMonth <= 0 || fiscalMonth > 13) {
			throw new FiscalMonthRangeException(fiscalMonth);
		}
		_fiscalYear = fiscalYear;
		_fiscalMonth = fiscalMonth;
	}

	@Override
	public String toString() {
		if (_fiscalMonth < 10) {
			return String.format("%d_0%d", _fiscalYear, _fiscalMonth);
		}
		return String.format("%d_%d", _fiscalYear, _fiscalMonth);
	}

	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof MonthIdentity)) {
			return false;
		}

		MonthIdentity monthId = (MonthIdentity) obj;
		return monthId._fiscalMonth == this._fiscalMonth
				&& monthId._fiscalYear == this._fiscalYear;
	}

	@Override
	public int hashCode() {
		return _fiscalYear * 100 + _fiscalMonth;
	}

	/**
	 * 
	 * @param str
	 * @return
	 * @throws MonthIdentityFormatException
	 */
	public static MonthIdentity parse(String str)
			throws MonthIdentityFormatException {
		try {
			String yearStr = str.substring(0, 4);
			String monthStr = str.substring(5, 7);
			int year = Integer.parseInt(yearStr);
			int month = Integer.parseInt(monthStr);

			MonthIdentity id = new MonthIdentity(year, month);

			return id;
		} catch (FiscalYearRangeException e) {
			throw new MonthIdentityFormatException();
		} catch (FiscalMonthRangeException e) {
			throw new MonthIdentityFormatException();
		} catch (IndexOutOfBoundsException e) {
			throw new MonthIdentityFormatException();
		} catch (NumberFormatException e) {
			throw new MonthIdentityFormatException();
		}
	}

	/**
	 * 
	 * @param monthId
	 * @return the month count between the month identity. 1 when this bigger
	 *         than monthId, 0 when equal, -1 when smaller
	 */
	public int compareTo(MonthIdentity monthId) {
		int ret = (this._fiscalYear - monthId._fiscalYear) * 12
				+ this._fiscalMonth - monthId._fiscalMonth;
		return ret;
	}

	/**
	 * add month
	 */
	public MonthIdentity addMonth() {
		int month = _fiscalMonth + 1;
		int year = _fiscalYear;
		if (month > 12) {
			month = 1;
			year++;
		}

		MonthIdentity newId = null;
		try {
			newId = new MonthIdentity(year, month);
		} catch (FiscalYearRangeException e) {
		} catch (FiscalMonthRangeException e) {
		}

		return newId;
	}
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Transaction/MonthIdentity.cs
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WinStore_FamilyAccountingCore.Transaction
{
    public class MonthIdentity : IComparable<MonthIdentity>
    {

        private readonly int _fiscalYear;
        public int FiscalYear { get { return _fiscalYear; } }

        private readonly int _fiscalMonth;
        public int FiscalMonth { get { return _fiscalMonth; } }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="fiscalYear"></param>
        /// <param name="fiscalMonth"></param>
        /// <exception cref="FiscalYearRangeException">Value of fiscal year</exception>
        /// <exception cref="FiscalMonthRangeException">Value of fiscal month</exception>
        public MonthIdentity(int fiscalYear, int fiscalMonth)
        {
            if (fiscalYear < 1000 || fiscalYear > 9999)
            {
                throw new FiscalYearRangeException(fiscalYear);
            }
            if (fiscalMonth <= 0 || fiscalMonth > 13)
            {
                throw new FiscalMonthRangeException(fiscalMonth);
            }
            _fiscalYear = fiscalYear;
            _fiscalMonth = fiscalMonth;
        }

        /// <summary>
        /// to string
        /// </summary>
        /// <returns></returns>
        public override String ToString()
        {
            if (_fiscalMonth < 10)
            {
                return String.Format("{0}_0{1}", _fiscalYear, _fiscalMonth);
            }
            return String.Format("{0}_{1}", _fiscalYear, _fiscalMonth);
        }

        /// <summary>
        /// equals
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(Object obj)
        {
            MonthIdentity monthId = obj as MonthIdentity;
            if (monthId == null)
            {// not match type MonthIdentity
                return false;
            }
            return monthId._fiscalMonth == this._fiscalMonth
                    && monthId._fiscalYear == this._fiscalYear;
        }

        /// <summary>
        /// Get hash code
        /// </summary>
        /// <returns></returns>
        public override int GetHashCode()
        {
            return _fiscalYear * 100 + _fiscalMonth;
        }


        /// <summary>
        /// parse string to month identity
        /// </summary>
        /// <param name="str"></param>
        /// <returns></returns>
        /// <exception cref="MonthIdentityFormatException">input string with error month identity format</exception>
        public static MonthIdentity Parse(String str)
        {
            try
            {
                string yearStr = str.Substring(0, 4);
                string monthStr = str.Substring(5, 2);
                int year = Int32.Parse(yearStr);
                int month = Int32.Parse(monthStr);

                MonthIdentity id = new MonthIdentity(year, month);

                return id;
            }
            catch (ArgumentOutOfRangeException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
            catch (FiscalYearRangeException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
            catch (FiscalMonthRangeException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
            catch (ArgumentNullException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
            catch (FormatException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
            catch (OverflowException e)
            {
                throw new MonthIdentityFormatException(e.Message);
            }
        }


        /// <summary>
        /// Add month, for example, current month identity is 2012_08, return is 2012_09
        /// </summary>
        /// <returns></returns>
        public MonthIdentity AddMonth()
        {
            int month = _fiscalMonth + 1;
            int year = _fiscalYear;
            if (month > 12)
            {
                month = 1;
                year++;
            }

            MonthIdentity newId = null;
            try
            {
                newId = new MonthIdentity(year, month);
            }
            catch (FiscalYearRangeException e)
            {
                throw new SystemException(e); // bug
            }
            catch (FiscalMonthRangeException e)
            {
                throw new SystemException(e); // bug
            }

            return newId;
        }

        /// <summary>
        /// Comparable
        /// </summary>
        /// <param name="other"></param>
        /// <returns></returns>
        public int CompareTo(MonthIdentity other)
        {
            int ret = (this._fiscalYear - other._fiscalYear) * 12
                    + this._fiscalMonth - other._fiscalMonth;
            return ret;
        }
    }

}
